### ISSUE TITLE
guest additions fails with VB 5.1.22, part 2

### ISSUE THREAD

**Issue by rulatir** [dotless-de/vagrant-vbguest#252]
I am very sorry for doing this but I have no choice. I am essentially reposting issue #251 that was self-closed by the submitter. I am bitten by the same problem, it completely halts all my work, @HenrikBechmann 's solution doesn't work for me, and I fear nobody will notice my activity in a closed issue.

```
Verifying archive integrity... All good.
Uncompressing VirtualBox 5.1.22 Guest Additions for Linux...........
VirtualBox Guest Additions installer
You appear to have a version of the VirtualBox Guest Additions
on your system which was installed from a different source or using a
different type of installer.  If you installed it from a package from your
Linux distribution or if it is a default part of the system then we strongly
recommend that you cancel this installation and remove it properly before
installing this version.  If this is simply an older or a damaged
installation you may safely proceed.

Do you wish to continue anyway? [yes or no]

Cancelling installation.
```

**Comment by rulatir** [comment]
I managed to resovle the issue by installing manually and replying "yes" to the question. I think the user of the vbguest plugin should be offered this choice instead of the plugin just bailing out.

**Comment by tswestendorp** [comment]
same issue here..

**Comment by HenrikBechmann** [comment]
I believe the issue was that the ['fix' for 5.1.21](https://github.com/dotless-de/vagrant-vbguest/issues/251#issuecomment-298205893) rendered /sbin/mount.vboxsf unable to be destroyed on uninstall of 5.1.21 to prepare for 5.1.22. So the solution is to manually destroy /sbin/mount.vboxsf upon encountering the above error, or avoid the whole thing by uninstalling 5.1.21 guest additions manually before upgrading to 5.1.22. Uninstall can be done by going to /opt/VBoxGuestAdditions-5.1.21/ and running `sudo ./uninstall.sh`

**Comment by rulatir** [comment]
No, that was not the issue in my case, as I *have* deleted the symlink and still received the same message upon retrying. I kept receiving it, and I thought "what if I were able to answer 'yes' to the question?". So I performed a manual installation in the guest, answered "yes", and it works fine.

**Comment by fnordfish** [comment]
@rulatir I'm sorry, but vbguest cannot communicate with interactive programs. This means there is no way to read your answer to the `Do you wish to continue anyway? [yes or no]` question and relay it the program. (I'm happy to provide more details on why this is not possible if you like.)
So, you'd need to avoid anything that requires manual user input. The `VBoxLinuxAdditions.run` tool is pretty opaque, but there is a `force` flag which should to what you what. You can set it like this:

```ruby
config.vbguest.installer_arguments = "--nox11 --force"
```

if this does not trigger the force, try:

```ruby
config.vbguest.installer_arguments = "--nox11 -- --force"
```

(Again, the `VBoxLinuxAdditions.run`'s  documentation is quite lacking)


However, the best way of dealing with guest additions not installed via the iso is to uninstall them. In most cases they come via some package management.

**Comment by esmerino** [comment]
@fnordfish Just ty like it:

`config.vbguest.installer_arguments = %w{--nox11 -- --force}`

**Comment by benh57** [comment]
Same problem here. Perhaps vbguest could detect this state and rerun itself with those arguments above.

**Comment by benh57** [comment]
Looks like I got this error even when upgrading from a much older version on a vbox i hadn't used in a long time. Fortunately the --force argument works as above.

Copy iso file C:\Program Files\Oracle\VirtualBox\VBoxGuestAdditions.iso into the box /tmp/VBoxGuestAdditions.iso
mount: /dev/loop0 is write-protected, mounting read-only
Installing Virtualbox Guest Additions 5.1.22 - guest version is 4.3.15
Verifying archive integrity... All good.
Uncompressing VirtualBox 5.1.22 Guest Additions for Linux...........
VirtualBox Guest Additions installer
You appear to have a version of the VirtualBox Guest Additions
on your system which was installed from a different source or using a
different type of installer.  If you installed it from a package from your
....

**Comment by ayen-tibco** [comment]
+1 
Version 0.14.2 plugin tries to install VirtualBoxAddition 5.1.20 continually while my Virtualbox has been upgraded to 5.1.22. The guest was originally running VBoxGuestAddition 4.2.30.  I had to manually remove the old VBoxGuestAddition and install the new one since the plugin didn't work.

**Comment by francoisjacques** [comment]
As stated in the README (rtfm 😉 ), `vbguest.installer_arguments` accepts an array. With the following syntax, the fix proposed by @fnordfish works great for me.

```
      config.vbguest.installer_arguments = ['--nox11', '-- ', '--force']
```

**Comment by Shalamus** [comment]
Inside the VM I removed Vbox* directories under 
/opt/
then I reinstalled VM guest addition using the following command

`vagrant vbguest --do install <VM name>`

**Comment by c-sonntag** [comment]
Hey everyone!
I have solved this problem with custom class vbguest:installer, inserted in the configuration.
I know, I can directly patch the "vbguest" package, but I'm not an expert in Ruby.
So, in your configuration part, add code like’s it:
```
VAGRANTFILE_API_VERSION = "2"
Vagrant.configure(VAGRANTFILE_API_VERSION) do |config|

  # [...]

  # -- Patch vbguest --
    class MyVbguestInstaller < VagrantVbguest::Installers::Linux
      def yes
        return 'echo yes\n | '
      end
    end
    config.vbguest.installer = MyVbguestInstaller
  # -- end : Patch vbguest --

  # [...]

end
```
I hope I've helped you.
Christophe.

**Comment by davidjb** [comment]
@c-sonntag Your solution is implemented in the core of the plugin now (fix released in 0.15.1); by default, the `yes` command is piped to the installer.  No extra options or `installer_arguments` required.
